### ISSUE TITLE
harvest Stack Exchange Q&A data

### ISSUE THREAD

**Issue by clstaudt** [LAION-AI/Open-Assistant#191]
https://stackexchange.com/sites

After reading a discussion here about how to build a model that detects "instruction-like" conversations on Twitter, I was wondering about data sources that are in instruction format by definition. I found no discussion on Stack Exchange in this repo. So this thread is to explore the possibility of using data from the numerous Stack Exchange sites.

**Comment by yk** [comment]
good idea, could be neat to get data from there. keep in mind that the main goal is task diversity, so we'd need to focus on that when scraping any external source, but SE in itself is already very diverse, so pretty promising.

**Comment by matan1905** [comment]
before scraping, there are some online dumps of this that might prove useful.
here is an example: https://www.kaggle.com/datasets/stackoverflow/stackoverflow?select=posts_answers

**Comment by clstaudt** [comment]
Also have a look at https://data.stackexchange.com before scraping.

**Comment by billray0259** [comment]
Specifically, what information do we want from the StackExchange websites? I assume at least:
- question title
- question body
- question votes
- list of
     - answer body
     - answer vote
 
Are there other pieces of information that should be collected?

Is the idea that OpenAssistant would be trained to produce the "answer body" conditioned on the "question title" + "question body?"

Perhaps the answers' votes could be used to train the reward model.

I'm curious to hear someone else's thoughts.

**Comment by getorca** [comment]
any thought given to only extracting excepted answers? Or just do what the pile did and run justext on the body? 

I might be able to jump on this? Anyone know how aggressive SE is on blocking crawlers, or can pull what we can from CC.

**Comment by getorca** [comment]
The pile also has a fairly large piece that's from SE. Do we think it'll be beneficial to have more in the training data? Anyways, probably a good idea to look how the SE data is formatted in the pile

**Comment by b-mc2** [comment]
I currently have a dataset I was putting together for this type of usage. 74 different Stack Exchange sources, post text, top rated answers, ratings, as well as extracted keywords. It's about 592k rows (152k questions,  440k answers) prefiltered by only questions with accepted answers and above a certain score, and I have it in parquet and JSONL with question/all answers on each line.

JSONL looks like this

![Image](https://user-images.githubusercontent.com/78936105/210173994-e7611b4b-4408-43d6-96e1-9a64b8c048ae.png)

**Comment by batelicm** [comment]
Stack exchange uploads it's dumps already

https://archive.org/details/stackexchange

That covers 367 sub stacks

Also of similar use is this reddit dump: https://files.pushshift.io/reddit/

**Comment by yk** [comment]
thanks @batelicm this is very valuable!

**Comment by smytjf11** [comment]
Hey all, happy to work on this issue. I can probably help with the Pile on the data side. I'll take a look at it a bit more and look for the things we've already talked about wanting to extract.

**Comment by clstaudt** [comment]
Is there some kind of "getting started" guide on how to make data sources like this compatible and useful for the project?

**Comment by yk** [comment]
> Is there some kind of "getting started" guide on how to make data sources like this compatible and useful for the project?

@lewtun do we already have that?

also, have a look at https://github.com/LAION-AI/Open-Assistant/pull/288 (not merged yet)

**Comment by b-mc2** [comment]
Added a notebook #355 (not merged yet) to ingest Stack Exchange from the data dumps. At the end I added the Open-Assistant Data Scheme from #288

**Comment by Kinvert** [comment]
https://download.kiwix.org/zim/stack_exchange/

This has a ton of stack exchanges. They are in zim format which works with Kiwix.

https://www.kiwix.org/

The 7z's linked above by @batelicm might work better but this could serve as a backup if that doesn't pan out.

Also note, you can go up a folder level to get access to more zim files from all over the place, beyond stack exchanges, to include sources such as Gutenberg, Wikibooks, and more.

https://download.kiwix.org/zim/

Another source to watch is Academic Torrents. It could help with sourcing more data.

**Comment by ontocord** [comment]
@smytjf11 re the pile, there are actually a lot of good things we can extract from there -  mathmatics instructions,  coding instructions, patent title to summary and summary to title (you could simulate a conversation about writing a patent by parsing the summaries into sentences. they are often written as the main parts of the patents). ping in the discord and we can discuss more.

**Comment by ontocord** [comment]
> Is there some kind of "getting started" guide on how to make data sources like this compatible and useful for the project?

we don't have a format yet, and at this stage it's more important to gather the data. my format as been [instruction->response, instruction->response, ...]  if possible, we would also like at the end something like [instruction -> [(response1, score1), (response2, score2), etc]

**Comment by ontocord** [comment]
> Added a notebook #355 (not merged yet) to ingest Stack Exchange from the data dumps. At the end I added the Open-Assistant Data Scheme from #288

cc @Vechtomov. @b-mc2 have we exported stack exchange into dialog format yet on hf?

**Comment by TommasoBendinelli** [comment]
Do you think that using the score associated to each answer as reward signal for RL would be a good idea? Or is it better to use this data for finetuning?

(Hello, first post here, great project!)

**Comment by bitplane** [comment]
> Also have a look at https://data.stackexchange.com before scraping.

Before messing with the bigdquery guff, command line hackers might enjoy the raw dumps:

https://archive.org/download/stackexchange

@TommasoBendinelli 

I think both maybe? I guess put the score in the metadata.

**Comment by Free-Radical** [comment]
Howdy, first time contributer here. I've done some scraping and data manipulation before. Pretty decet with Python. Somebody pls tell me how i begin. Thanks!

**Comment by stefnnn** [comment]
Hey @Free-Radical. It looks like a notebook to convert StackExchange data to Open Assistant format was already created by @b-mc2  and merged, see here: https://github.com/LAION-AI/Open-Assistant/pull/355.

I think the next step would be to discuss, whether there are benefits in further massaging the data, then it will need to be converted into parquet files and uploaded to HF (see https://projects.laion.ai/Open-Assistant/docs/data/datasets).

**Comment by Free-Radical** [comment]
cool, pls tell me next steps. Or if you think you can use me elsewhere pls let me know, also how do u guys do DM's when necessary? PS where could i see a review of the whole project?

**Comment by stefnnn** [comment]
I gave it a go and after some fights with 100GB XML files I have prepared a PR https://github.com/LAION-AI/Open-Assistant/pull/2848
